### PR TITLE
feat: use orjson to serialize responses

### PR DESCRIPTION
**Pull request by ryanfang5**

### CODE REVIEW

**Comment by ryanfang5** [comment]
Used version 2.33.6 of virtool-core as the latest version causes a pydantic validation error which I believe should be resolved in another commit.

Uses orjson to serialize json responses, however extra methods that return strings needed to be made as SQLAlchemy expects strings.

It was possible to get rid of `CustomEncoder`- `isoformat` could also be replaced but I am not sure if it is necessary.

File: virtool/api/custom_json.py [hunk 1]
```diff
@@ -44,28 +25,59 @@ def isoformat(obj: datetime.datetime) -> str:
     return obj.replace(tzinfo=datetime.timezone.utc).isoformat().replace("+00:00", "Z")
 
 
-def dumps(obj: object) -> str:
+def default(obj):
```
Comment (igboyes) [inline virtool/api/custom_json.py:28]: Write a docstring that describes where this fits in the grand scheme.

File: virtool/api/custom_json.py [hunk 2]
```diff
@@ -44,28 +25,59 @@ def isoformat(obj: datetime.datetime) -> str:
...
+    )
 
 
-def pretty_dumps(obj: object) -> str:
+def pretty_dumps(obj: object) -> bytes:
     """
     A wrapper for :func:`json.dumps` that applies pretty formatting to the output.
 
     Sorts keys and adds indentation. Used as ``dumps`` argument for
     :func:`.json_response`.
```
Comment (igboyes) [inline virtool/api/custom_json.py:56]: This docstring needs an update.

File: virtool/api/custom_json.py [hunk 3]
```diff
@@ -44,28 +25,59 @@ def isoformat(obj: datetime.datetime) -> str:
...
+    """
+    return pretty_dumps(obj).decode()
+
+
+def orjson_serializer(obj) -> str:
+    """
+    Used by SQLAlchemy as they expect strings.
     """
-    return json.dumps(obj, cls=CustomEncoder, indent=4, sort_keys=True)
+    return dumps(obj).decode()
```
Comment (igboyes) [inline virtool/api/custom_json.py:83]: Can you explain why there are two of these?

File: virtool/indexes/api.py [hunk 4]
```diff
@@ -155,7 +154,7 @@ async def download_otus_json(req):
             db, req.app["config"], index["manifest"]
         )
 
-        json_string = json.dumps(patched_otus, cls=CustomEncoder)
+        json_string = dumps(patched_otus)
 
         await run_in_thread(compress_json_with_gzip, json_string, json_path)
```
Comment (igboyes) [inline virtool/indexes/api.py:159]: May as well.
```suggestion
        await run_in_thread(compress_json_with_gzip, dumps(patched_otus), json_path)
```

File: virtool/indexes/tasks.py [hunk 5]
```diff
@@ -89,7 +89,7 @@ async def _add_one_json(self, index: Document):
 
             await run_in_thread(
                 compress_json_with_gzip,
-                json.dumps(
+                orjson.dumps(
```
Comment (igboyes) [inline virtool/indexes/tasks.py:92]: Just use your custom `dumps`. What is `"targets"` gets a `datetime` in it at some point.

File: virtool/pg/utils.py [hunk 6]
```diff
@@ -36,7 +37,8 @@ async def connect(postgres_connection_string: str) -> AsyncEngine:
     try:
         pg = create_async_engine(
             postgres_connection_string,
-            json_serializer=pretty_dumps,
+            json_serializer=pretty_orjson_serializer,
```
Comment (igboyes) [inline virtool/pg/utils.py:40]: Why pretty?

File: virtool/api/custom_json.py [hunk 7]
```diff
@@ -9,30 +9,11 @@
...
-        if isinstance(obj, datetime.datetime):
-            return isoformat(obj)
-
-        if issubclass(type(obj), BaseModel):
-            return obj.dict(by_alias=True)
-
-        return json.JSONEncoder.default(self, obj)
-
-
 def isoformat(obj: datetime.datetime) -> str:
```
Comment (igboyes) [inline virtool/api/custom_json.py:17]: Are we still using this?

**Review by igboyes — CHANGES_REQUESTED** [review]

File: virtool/api/custom_json.py (hunk 3 above)
Comment (ryanfang5) [reply virtool/api/custom_json.py:83]: In pg/utils, the json_serializer was `pretty_dumps`. However, in pg/testing the serializer was `dumps`, so I made two to preserve that functionality.

File: virtool/pg/utils.py (hunk 6 above)
Comment (ryanfang5) [reply virtool/pg/utils.py:40]: I will change this to regular dumps and get rid of `pretty_orjson_serializer`.

File: virtool/api/custom_json.py (hunk 3 above)
Comment (igboyes) [reply virtool/api/custom_json.py]: That seems backwards to me.

I think we should get rid of the pretty serializer. The use I see for pretty is that the JSONB looks good in a snapshot or during debugging. There are better ways to do that than storing pretty JSON in Postgres.

Agree?

File: virtool/api/custom_json.py (hunk 7 above)
Comment (igboyes) [reply virtool/api/custom_json.py:17]: Ah. Thanks for that.

**Pull request merged**